### DISCUSSION TITLE
Combine zero-shot classification with keywords list

### DISCUSSION THREAD

**Discussion by mzhadigerov** [MaartenGr/BERTopic#2050] [Q&A] [answered]
Let's say I've created some custom topics:

```
["customer service", "chairman's statements", "delivery issues"]
```

Then I've "assigned" some keywords to those topics:

```
[
   {
      "customer service": ["smile", "tips", "service", "positive"],
      "chairman's statements": [...]
      "delivery issues": [...]
    }
]
```

Is there a way to model this kind of relationship in BERTopic? So maybe the model will consider the keywords assigned to each topic somehow? Something like a combo of `zeroshot_topic_list` and `seed_topic_list`, I guess. 

One way would be to just join the keywords and append them to the corresponding topics, e.g.:

```
"customer service": ["smile", "tips", "service", "positive"]  --->  "customer service: smile, tips, service, positive" 
```

or the other way around is to add topics as a new keyword:

```
"customer service": ["smile", "tips", "service", "positive"]  ----> ["customer service", "smile", "tips", "service", "positive"]
```

But both approaches seem weird to me...

Just curious that's possible. Thanks!

**Comment by MaartenGr** [comment]
Hmmm, there is indeed no such option with zero-shot topic modeling at the moment but you can [seed words](https://maartengr.github.io/BERTopic/getting_started/seed_words/seed_words.html) instead. It allows you to increase the potential weight of certain potential keywords that might end up in your topic representations. It would, however, not be specific to those topics and increase their weights across all topics.

**Reply by mzhadigerov** [reply]
@MaartenGr Yes, what I did so far was:

Used `zeroshot_topic_list` where I combined topic name and keywords:

```
"customer service": ["smile", "tips", "service", "positive"]  --->  "customer service: smile, tips, service, positive" 
```

and added those keywords in `seed_words`.

But as you wrote, it is not exactly what I'm looking for, because seed words increases weights of the keywords across all topics. However, I think that adding such a feature would be great, because in many cases domain experts might know the topics in the documents (`already handled by zeroshot`) and what words/keywords are associated with those topics. 

So the keywords would be kind of extra "force" to converge the documents towards certain topics. As in my experience, domain experts (at least in financial sector) are not very good at choosing topics. They tend to choose something like (`financial analysis`, `gross margin`, `sustainability analysis`) etc.

You see, unlike in broader areas, like research papers, history books etc. (where documents are pretty well distinguishable by using cosine similarity between embeddings), in more niche areas (like finance), it is more beneficial to use combinations of keywords and topics together.

Otherwise, thanks for a wonderful library! It quickly became one of my favourites for Topic Modelling.

**Reply by MaartenGr** [reply, marked as answer]
> However, I think that adding such a feature would be great, because in many cases domain experts might know the topics in the documents (already handled by zeroshot) and what words/keywords are associated with those topics.

Just to be sure I understand correctly, do you want the keywords to affect the **topic assignment** or the **topic representation**?

If it is the **topic assignment**, then your approach of concatenating the zero-shot label and keywords is actually quite reasonable! Embedding models typically work by merging word embeddings so adding them here makes sense to me. You could make this more complex by embedding the keywords and documents separately and then applying cosine similarity but I think the result will remain similar. 

If it is the **topic representation**, then it might make sense to extend the functionality of `seed_words` to include the zero-shot topics. Having said that, I do think that even though `seed_words` work on a global level the actual calculation of c-TF-IDF values should make it so that small values remain small. Multiplying a very small c-TF-IDF value by 2 will still keep it small. 

> So the keywords would be kind of extra "force" to converge the documents towards certain topics.

Another option would be to train a separate TF-IDF model (or even just BoW) on the documents and use cosine similarity between the documents and keywords. That would, however, introduce an additional layer of complexity which might not be needed.

**Reply by mzhadigerov** [reply]
Topic representation triggers after documents been clustered, right? In that case, I want keywords and custom topics to be assigned during ***topic assignment***. Alright, then I will stick to my approach. Thank you!
